### PR TITLE
Allow streaming from the /chat/completions route

### PR DESCRIPTION
**Pull request by arcticfly**
We now allow streaming from our /chat/completions route for base and fine-tuned OpenAI models. We don't yet support streaming for fine-tuned Mistral models, but that should be coming shortly.

Additionally, this PR adds support for logging calls as part of the request to /chat/completions, instead of requiring an additional call to the /report route.

### CODE REVIEW

File: app/src/server/api/external/v1Api.router.ts [hunk 1]
```diff
@@ -105,8 +93,8 @@ export const v1ApiRouter = createOpenApiRouter({
           .optional(),
       }),
     )
-    .output(chatCompletionOutput)
-    .mutation(async ({ input, ctx }): Promise<ChatCompletion> => {
+    .output(z.union([chatCompletionOutput.nullable(), z.any()]))
```
Comment (arcticfly) [inline app/src/server/api/external/v1Api.router.ts:97]: I'm not sure of the right way to type a ReadableStream in zod. z.any works in the short term, but we should come up with a better solution.

Comment (corbt) [reply app/src/server/api/external/v1Api.router.ts:97]: I don't think it's possible. Pretty sure Zod is just focused on typing things that are JSON-serializable, which a readable stream definitely isn't.

File: app/src/server/api/external/v1Api.router.ts [hunk 2]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
+        }
+      }
 
-        return completion;
-      } catch (error: unknown) {
-        console.error(error);
-        throw new TRPCError({
-          message: `Failed to get completion: ${(error as Error).message}`,
-          code: "BAD_REQUEST",
+      const logRequest = ctx.headers["op-log-request"] === "true";
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts]: I think we should default to reporting unless you explicitly opt out. That makes it much simpler for us to tell someone "if you want to start using openpipe with [your random openai wrapper library], just use our baseUrl" instead of them also having to figure out how to set a custom header with their wrapper lib, which may not even be possible.

To avoid double-reporting from our SDKs in the short-term we can only default to `true` if you're calling a non-openpipe model. But we should also update our SDKs to set this to `true` and stop calling the reporting endpoint separately.

File: app/src/server/api/external/v1Api.router.ts [hunk 3]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
+          tags = JSON.parse(ctx.headers["op-tags"] as string);
+        } catch (error: unknown) {
+          // ignore
+        }
+      }
+
+      if (completion instanceof Stream) {
+        // split stream to avoid locking the read mutex
+        const [recordStream, outputStream] = completion.tee();
+        void captureRecordUsage({
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts]: Shouldn't this go inside `recordUsage`?

File: app/src/server/api/external/v1Api.router.ts [hunk 4]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
-      } catch (error: unknown) {
-        console.error(error);
-        throw new TRPCError({
-          message: `Failed to get completion: ${(error as Error).message}`,
-          code: "BAD_REQUEST",
+      const logRequest = ctx.headers["op-log-request"] === "true";
+      let tags: Record<string, string> = {};
+      if (ctx.headers["op-tags"]) {
+        try {
+          tags = JSON.parse(ctx.headers["op-tags"] as string);
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts:162]: Let's actually just throw an error if you pass in an `op-tags` instance but it isn't formatted correctly. That's better than silently swallowing it and the user not knowing if they set the wrong header name, or misformatted it, or something else.

File: app/src/server/api/external/v1Api.router.ts [hunk 5]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
+          inputPayload,
+          completion: recordStream,
+          logRequest,
+          fineTune,
+          tags,
         });
+        return outputStream.toReadableStream();
       }
+
+      void recordUsage({
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts]: Let's add a `.catch(e => captureException(e))` or similar to all of these so see in Sentry if these are breaking. We should strive to never silently swallow errors.

File: app/src/server/api/external/v1Api.router.ts [hunk 6]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
+        try {
+          tags = JSON.parse(ctx.headers["op-tags"] as string);
+        } catch (error: unknown) {
+          // ignore
+        }
+      }
+
+      if (completion instanceof Stream) {
+        // split stream to avoid locking the read mutex
+        const [recordStream, outputStream] = completion.tee();
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts:175]: Nice, I should use `tee` more.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts [hunk 7]
```diff
@@ -34,27 +36,42 @@ export async function getCompletion2(
...
-        completion.choices[0].message,
-      ) as ChatCompletionMessage;
-    }
     return completion;
-  } else {
-    return getModalCompletion(fineTune, prunedInput);
   }
+
+  // TODO: create pipeline without this conversion once OpenAI supports tool_calls for their fine-tuned models
+  input = convertToolCallInputToFunctionInput(input);
```
Comment (corbt) [inline app/src/modelProviders/fine-tuned/getCompletion-2.ts]: I'm a little confused about what's going on here. Why are we converting tool calls only for non-streamed responses? Shouldn't tool calls be treated the same way independent of whether you're streaming or not?

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts [hunk 8]
```diff
@@ -21,7 +23,7 @@ import { type TypedFineTune } from "~/types/dbColumns.types";
 export async function getCompletion2(
```
Comment (corbt) [inline app/src/modelProviders/fine-tuned/getCompletion-2.ts:25]: So I realize it looks ugly but maybe we want to add multiple function declarations here? That way typescript knows that if we pass in stream: true, we get a stream, and if not not. Eg:

```ts
export async function getCompletion2(
  fineTune: TypedFineTune,
  input: ChatCompletionCreateParamsNonStreaming,
): Promise<ChatCompletion>;
export async function getCompletion2(
  fineTune: TypedFineTune,
  input: ChatCompletionCreateParamsStreaming,
): Promise<Stream<ChatCompletionChunk>>;
export async function getCompletion2(
  fineTune: TypedFineTune,
  input: ChatCompletionCreateParams,
): Promise<ChatCompletion | Stream<ChatCompletionChunk>>;
```

File: app/src/server/api/external/v1Api.router.ts [hunk 9]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
-        where: { slug: modelSlug },
-      });
-
-      if (!fineTune) {
-        throw new TRPCError({ message: "The model does not exist", code: "NOT_FOUND" });
-      }
+      const fineTune =
+        (await prisma.fineTune.findUnique({
+          where: { slug: modelSlug },
+        })) ?? undefined;
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts]: What's the `?? undefined` for?

File: app/src/server/api/external/v1Api.router.ts [hunk 10]
```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
...
 
-      const typedFT = typedFineTune(fineTune);
-
-      if (typedFT.projectId !== key.projectId) {
+      if (inputPayload.model.startsWith("openpipe:") && !fineTune) {
```
Comment (corbt) [inline app/src/server/api/external/v1Api.router.ts]: The flow of this function feels somewhat hard to follow with me. I think it would be cleaner to just have an `if (inputPayload.model.startswith("openpipe:"))` at the top and split at that point, instead of having some checks here and then a separate `if (fineTune)` down below.

File: app/src/server/tasks/generateTestSetEntry.task.ts [hunk 11]
```diff
@@ -125,21 +125,22 @@ export const generateTestSetEntry = defineTask<GenerateTestSetEntryJob>({
...
         response_format: datasetEntry.response_format ?? undefined,
+        stream: false,
       };
 
       try {
         if (isComparisonModel(modelId)) {
           completion = await getOpenaiCompletion(rawDatasetEntry.dataset.projectId, input);
         } else if (fineTune) {
-          completion = await getCompletion2(fineTune, input);
+          completion = (await getCompletion2(fineTune, input)) as ChatCompletion;
```
Comment (corbt) [inline app/src/server/tasks/generateTestSetEntry.task.ts]: If we make the change to the typings of `getCompletion2` we can drop this. If we don't though should make this an actual check on the return type, not a type assertion. Too easy to make make mistakes with type assertions!

File: app/src/server/api/external/v1Api.router.ts (hunk 4 above)
Comment (corbt) [reply app/src/server/api/external/v1Api.router.ts:162]: Also we need to use zod to validate this is actually a record of strings.

File: app/src/utils/recordRequest.ts [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+import { posthogServerClient } from "./analytics/serverAnalytics";
+import {
+  countLlamaInputTokens,
+  countLlamaOutputTokens,
+  countOpenAIChatTokens,
+} from "./countTokens";
+import { default as openAIModelProvider } from "~/modelProviders/openai-ChatCompletion";
+import { type BaseModel } from "~/server/fineTuningProviders/types";
+
+export const captureRecordUsage = async ({
```
Comment (corbt) [inline app/src/utils/recordRequest.ts]: This is basically a wrapper for `recordUsage` that works with streaming, right? Can't we just let `recordUsage` take a completion of type stream or chatcompletion and handle this internally? A separate wrapper I feel like just adds indirection for not much gain in this case.

File: app/src/server/api/external/v1Api.router.ts (hunk 5 above)
Comment (corbt) [reply app/src/server/api/external/v1Api.router.ts]: Also let's put this inside an `else`, so it's clear that this is stuff that happens if we're *not* streaming. I missed that on my first read-through and got confused.

File: app/src/server/api/external/v1Api.router.ts (hunk 2 above)
Comment (arcticfly) [reply app/src/server/api/external/v1Api.router.ts]: Sure. I think that should go in a separate PR as a larger refactor of our SDKs to allow streaming (currently our SDKs will break if you try to use the streaming features this PR enables for fine-tuned models).

File: app/src/server/api/external/v1Api.router.ts (hunk 3 above)
Comment (arcticfly) [reply app/src/server/api/external/v1Api.router.ts]: I don't see a problem with combining recordUsage and captureRecordUsage. Let's do it.

File: app/src/server/api/external/v1Api.router.ts (hunk 9 above)
Comment (arcticfly) [reply app/src/server/api/external/v1Api.router.ts]: Setting it to undefined makes our calls to `recordUsage` a little cleaner. Hopefully the refactor clarified its purpose.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts (hunk 7 above)
Comment (arcticfly) [reply app/src/modelProviders/fine-tuned/getCompletion-2.ts]: I added a comment earlier in the file to clarify. Basically converting a streamed function call into a tool call output on the fly would be complicated.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts (hunk 8 above)
Comment (arcticfly) [reply app/src/modelProviders/fine-tuned/getCompletion-2.ts:25]: It's not ugly. It's beautiful.

File: app/src/server/api/external/v1Api.router.ts (hunk 10 above)
Comment (arcticfly) [reply app/src/server/api/external/v1Api.router.ts]: I refactored it

**Review by corbt — APPROVED** [review]
great work here.
